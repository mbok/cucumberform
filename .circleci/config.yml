version: 2
jobs:
  build:
    machine: true
    steps:
      - checkout
      # build the application image
      - run:
          name: Build and test
          command: docker build -t stephub/build:$CIRCLE_BRANCH .
      - run:
          name: Dockerize CLI app
          command: docker build --build-arg "BUILD_IMAGE=stephub/build:$CIRCLE_BRANCH" --build-arg "BUILD_VERSION=$CIRCLE_BRANCH" -t stephub/cli:$CIRCLE_BRANCH ./cli