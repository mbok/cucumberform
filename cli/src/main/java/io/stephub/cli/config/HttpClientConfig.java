package io.stephub.cli.config;

import lombok.extern.slf4j.Slf4j;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.info.BuildProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
@Slf4j
public class HttpClientConfig {
    @Autowired(required = false)
    private BuildProperties buildProperties;

    @Bean
    public OkHttpClient.Builder httpClientBuilder() {
        return new OkHttpClient.Builder().addInterceptor(new HttpLoggingInterceptor(
                log::debug
        ).setLevel(HttpLoggingInterceptor.Level.BASIC))
                .addInterceptor(chain -> {
                    final Request originalRequest = chain.request();
                    final Request requestWithUserAgent = originalRequest.newBuilder()
                            .header("User-Agent", this.getUserAgent())
                            .build();
                    return chain.proceed(requestWithUserAgent);
                });
    }

    private String getUserAgent() {
        final StringBuilder ua = new StringBuilder("stephub.cli.v");
        if (this.buildProperties != null) {
            ua.append(this.buildProperties.getVersion());
        }
        return ua.toString();
    }
}
