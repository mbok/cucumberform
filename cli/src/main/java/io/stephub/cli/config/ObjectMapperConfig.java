package io.stephub.cli.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import io.stephub.json.jackson.ObjectMapperConfigurer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;

import static io.stephub.json.jackson.ObjectMapperConfigurer.createObjectMapper;

@Configuration
public class ObjectMapperConfig {
    public static final String YAML = "yamlObjectMapper";

    @Primary
    @Bean
    public ObjectMapper objectMapper() {
        return createObjectMapper();
    }

    @Bean(value = YAML)
    public ObjectMapper yamlObjectMapper() {
        final ObjectMapper yamlMapper = new ObjectMapper(new YAMLFactory());
        ObjectMapperConfigurer.configure(yamlMapper);
        return yamlMapper;
    }
}
