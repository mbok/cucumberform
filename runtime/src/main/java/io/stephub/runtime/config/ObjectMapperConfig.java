package io.stephub.runtime.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.stephub.json.jackson.ObjectMapperConfigurer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;

import javax.annotation.PostConstruct;

@Configuration
public class ObjectMapperConfig {
    @Autowired
    private ObjectMapper objectMapper;

    @PostConstruct
    public void setUp() {
        ObjectMapperConfigurer.configure(this.objectMapper);
    }
}
