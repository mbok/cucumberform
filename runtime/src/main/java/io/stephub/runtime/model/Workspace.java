package io.stephub.runtime.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;

import javax.validation.Valid;
import java.util.ArrayList;
import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Data
@Builder
@EqualsAndHashCode
@ToString(of = {"id"})
public class Workspace {
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private String id;
    private String name;
    @Valid
    @Singular
    private List<ProviderSpec> providers = new ArrayList<>();
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private List<Error> errors;

    @Value
    @Builder
    public static class Error {
         String path;
        String message;
    }
}
