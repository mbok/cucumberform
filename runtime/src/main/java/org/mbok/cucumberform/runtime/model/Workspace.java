package org.mbok.cucumberform.runtime.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;

import javax.validation.Valid;
import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Data
@Builder
@EqualsAndHashCode
@ToString(of = {"id"})
public class Workspace {
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private String id;
    private String name;
    @Valid
    @Singular
    private List<ProviderSpec> providers;
}
