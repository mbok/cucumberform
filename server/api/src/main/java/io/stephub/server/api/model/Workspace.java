package io.stephub.server.api.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.stephub.provider.api.util.Patterns;
import io.stephub.server.api.model.customsteps.CustomStepContainer;
import io.stephub.server.api.model.customsteps.StepDefinition;
import io.stephub.server.api.model.gherkin.Feature;
import io.stephub.server.api.model.gherkin.Fixture;
import io.stephub.server.api.validation.IProviderValidator;
import lombok.*;
import org.springframework.validation.ObjectError;

import javax.validation.Valid;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


@AllArgsConstructor
@NoArgsConstructor
@Data
@Builder
@EqualsAndHashCode
@ToString(of = {"id"})
public class Workspace implements CustomStepContainer {
    private String id;
    @NotEmpty
    private String name;

    @Builder.Default
    @Valid
    @NotNull
    private GherkinPreferences gherkinPreferences = new GherkinPreferences();

    @Valid
    @Singular
    private List<@IProviderValidator.Valid ProviderSpec> providers = new ArrayList<>();

    @Singular
    @Valid
    private List<StepDefinition> stepDefinitions = new ArrayList<>();

    @Singular
    @Valid
    private Map<@Pattern(regexp = Patterns.ID_PATTERN_STR) String, Variable> variables = new HashMap<>();

    @Singular
    @Valid
    private List<Feature> features = new ArrayList<>();

    @Singular
    @Valid
    private List<Fixture> beforeFixtures = new ArrayList<>();

    @Singular
    @Valid
    private List<Fixture> afterFixtures = new ArrayList<>();

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private List<ObjectError> errors = null;

}
