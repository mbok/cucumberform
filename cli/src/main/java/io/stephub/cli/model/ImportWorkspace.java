package io.stephub.cli.model;

import com.fasterxml.jackson.annotation.JsonView;
import io.stephub.server.api.model.Workspace;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;

@Data
@EqualsAndHashCode(callSuper = true)
public class ImportWorkspace extends Workspace {
    public static class InfoView {

    }

    @JsonView(InfoView.class)
    private List<String> featureFiles = new ArrayList<>();
}
