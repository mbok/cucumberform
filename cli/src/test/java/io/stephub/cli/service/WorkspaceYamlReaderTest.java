package io.stephub.cli.service;

import io.stephub.cli.config.ObjectMapperConfig;
import io.stephub.server.api.model.Workspace;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.io.IOException;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {WorkspaceYamlReader.class, ObjectMapperConfig.class})
@Slf4j
class WorkspaceYamlReaderTest {
    @Autowired
    private WorkspaceYamlReader reader;

    @Test
    public void testSimple() throws IOException {
        final Workspace workspace = this.reader.read(new ClassPathResource("workspace-simple.yml"));
    }
}