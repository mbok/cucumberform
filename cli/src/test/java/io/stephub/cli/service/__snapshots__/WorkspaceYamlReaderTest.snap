io.stephub.cli.service.WorkspaceYamlReaderTest.testFeaturesFromFile=[
  "{\"gherkinPreferences\":{\"assignmentKeywords\":[\"- assigned to @ATTRIBUTE\"],\"stepKeywords\":[\"Given\",\"When\",\"Then\",\"But\"]},\"providers\":[],\"stepDefinitions\":[],\"variables\":{},\"features\":[],\"beforeFixtures\":[],\"afterFixtures\":[],\"featureFiles\":[\"/home/mbok/IdeaProjects/stephub4/cli/target/test-classes/features/feature1.feature\"]}"
]


io.stephub.cli.service.WorkspaceYamlReaderTest.testMerged=[
  "{\"name\":\"abc\",\"gherkinPreferences\":{\"assignmentKeywords\":[\"- assigned to @ATTRIBUTE\"],\"stepKeywords\":[\"Given\",\"When\",\"Then\",\"But\"]},\"providers\":[],\"stepDefinitions\":[{\"type\":\"basic\",\"spec\":{\"pattern\":\"My step\",\"patternType\":\"simple\",\"arguments\":[],\"payload\":\"none\"},\"stepDefinitions\":[],\"steps\":[\"Given I do something\",\"Then I expect something\"],\"id\":\"simpleMy step\"},{\"type\":\"basic\",\"spec\":{\"pattern\":\"My winning step\",\"patternType\":\"simple\",\"arguments\":[],\"payload\":\"none\"},\"stepDefinitions\":[],\"steps\":[\"Given I win\"],\"id\":\"simpleMy winning step\"}],\"variables\":{\"var1\":{\"value\":null,\"schema\":{\"type\":\"string\"},\"default\":null}},\"features\":[{\"name\":\"Feature 1\",\"scenarios\":[{\"name\":\"Scene 1\",\"steps\":[\"Given I do something\",\"Then I expect something\"],\"tags\":[],\"comments\":[]}],\"tags\":[],\"comments\":[]}],\"beforeFixtures\":[],\"afterFixtures\":[],\"featureFiles\":[]}"
]


io.stephub.cli.service.WorkspaceYamlReaderTest.testSimple=[
  "{\"name\":\"abc\",\"gherkinPreferences\":{\"assignmentKeywords\":[\"- assigned to @ATTRIBUTE\"],\"stepKeywords\":[\"Given\",\"When\",\"Then\",\"But\"]},\"providers\":[],\"stepDefinitions\":[{\"type\":\"basic\",\"spec\":{\"pattern\":\"My step\",\"patternType\":\"simple\",\"arguments\":[],\"payload\":\"none\"},\"stepDefinitions\":[],\"steps\":[\"Given I do something\",\"Then I expect something\"],\"id\":\"simpleMy step\"}],\"variables\":{\"var1\":{\"value\":null,\"schema\":{\"type\":\"string\"},\"default\":null}},\"features\":[{\"name\":\"Feature 1\",\"scenarios\":[{\"name\":\"Scene 1\",\"steps\":[\"Given I do something\",\"Then I expect something\"],\"tags\":[],\"comments\":[]}],\"tags\":[],\"comments\":[]}],\"beforeFixtures\":[],\"afterFixtures\":[],\"featureFiles\":[]}"
]